### ISSUE TITLE
Potential NullPointerException in `AggregateReferenceDeserializerModifier`

### ISSUE THREAD

**Issue by alienisty** [spring-projects/spring-data-rest#2033]
AggregateReferenceDeserializerModifier does not update the BeanDeserializerBuilder if there is already a mapping for the bean type.
The problem is that, in some cases, having a mapping for an entity does not mean that that entity has been already processed or will be processed by AssociationUriResolvingDeserializerModifier, causing the URI deserialization to fail.

Attached is a minimal demo applilcation that reproduces the problem.
[demo-2033.zip](https://github.com/spring-projects/spring-data-rest/files/6753767/demo-2033.zip)

**Comment by alienisty** [comment]
On further analysis, I found that propertyType at line AggregateReferenceDeserializerModifier:108 can be null. In such, case we would get a NullPointerException at line AggregateReferenceDeserializerModifier109, breaking the entire loop of customisations, leaving some of the entities not processed by AssociationUriResolvingDeserializerModifier.
To put in context in AggregateReferenceDeserializerModifier:

```java
public BeanDeserializerBuilder updateBuilder(DeserializationConfig config, BeanDescription beanDesc,
		BeanDeserializerBuilder builder) {

	// Type is aggregate itself, already handled by AssociationUriResolvingDeserializerModifier
	if (mappings.hasMappingFor(beanDesc.getBeanClass())) {
		return builder;
	}

	TypeInformation<?> type = ClassTypeInformation.from(beanDesc.getBeanClass());
	ValueInstantiatorCustomizer customizer = new ValueInstantiatorCustomizer(builder.getValueInstantiator(), config);
	Iterator<SettableBeanProperty> properties = builder.getProperties();

	while (properties.hasNext()) {

		SettableBeanProperty property = properties.next();

		TypeInformation<?> propertyType = type.getProperty(property.getName()); // <-- can be null and it will be for some cases
		TypeInformation<?> actualType = propertyType.getActualType(); // causing a NullPointerException here

		if (!mappings.exportsMappingFor(actualType.getType())) {
			continue;
		}

		UriStringDeserializer uriStringDeserializer = new UriStringDeserializer(actualType.getType(), converter);
		JsonDeserializer<?> deserializer = wrapIfCollection(propertyType, uriStringDeserializer, config);

		customizer.replacePropertyIfNeeded(builder, property.withValueDeserializer(deserializer));
	}

	return builder;
}
```

**Comment by odrotbohm** [comment]
I have added the `null` check easily, but have a hard time to construct a scenario in which that case actually comes up. Do you have a hint at what might be causing a property to be exposed on the builder but not available on the type? I suspected mixins first, but that doesn't seem to be it.

**Comment by mjustin** [comment]
@odrotbohm Do either of the issues I reported that were closed as duplicates help? #2037, #2038

**Comment by alienisty** [comment]
> I have added the `null` check easily, but have a hard time to construct a scenario in which that case actually comes up. Do you have a hint at what might be causing a property to be exposed on the builder but not available on the type? I suspected mixins first, but that doesn't seem to be it.

While running the provided demo, I can see, for example, that propertyType is null when analysing the "_embedded" property of CollectionModel and that happens because there is no "_embedded" property defined for such class.

**Comment by odrotbohm** [comment]
Never mind, I finally got my mixin setup for the unit test working. Fix coming in a bit.

**Comment by odrotbohm** [comment]
This should be fixed with e1dd82c and was back-ported into 3.5 and 3.4 scheduled for release by the end of the week.
